File: Java/src/com/dbc/algorithm/Daily/BestTimeToBuyAndSellStockWithTransactionFee.java
package com.dbc.algorithm.Daily;

public class BestTimeToBuyAndSellStockWithTransactionFee {
    public static int maxProfit(int[] prices, int fee) {
        int length = prices.length;
        if (length == 0)
            return 0;

        int buy = -prices[0], sell = 0;
        for (int i = 1; i < length; i++) {
            int temp = buy;
            buy = Math.max(buy, sell - prices[i]);
            sell = Math.max(sell, temp + prices[i] - fee);
        }

        return Math.max(0, sell);
    }
}

File: Python/best_time_to_buy_and_sell_stock_with_transaction_fee.py
def max_profit_dp(prices: list, fee: int) -> int:
    if not prices:
        return 0
    
    length = len(prices)
    buy, sell = -prices[0], 0
    for i in range(1, length):
        temp = buy
        buy = max(buy, sell - prices[i])
        sell = max(sell, temp + prices[i] - fee)
    
    return max(0, sell)

if __name__ == "__main__":
    print(max_profit_dp([1, 3, 2, 8, 4, 9], 2))